File object notice stream

iblanco11981870 opened the pull request:
<!-- Please confirm that your work is based on the correct branch. -->
<!-- Bug fixes should be based on *trunk-patch*. -->
<!-- Backwards compatible new features should be based on *trunk-minor*. -->
<!-- Incompatible API changes should be based on *trunk-major*. -->

## Description

<!-- Describe your changes in detail. -->
Provide a Python file-like object that writes to the HOOMD notice message stream.
## Motivation and context
Write a notice message to a file using `hoomd.Device`.
<!--- Why is this change required? What problem does it solve? -->
This change is useful for tracking error outputs when you are running operations in parallel or with multiple partitions. This will allow the user to save errors or outputs to each partition into a separate file. 
<!-- Replace ??? with the issue number that this pull request resolves. -->
Resolves #1449 
## How has this been tested?
We implemented the appropriate tests in ``test_device.py``
<!--- Please describe in detail how you tested your changes. -->
Tested outputs of different data types and notice levels.
<!--- Please build the sphinx documentation and check that any changes to
      documentation display properly. -->

## Change log

<!-- Propose a change log entry. -->
```
`hoomd.device.NoticeFile` provides a Python file-like object that can save notice message streams.
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.rst).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

iblanco11981870 committed:
changes into output notice file in hoomd

iblanco11981870 committed:
implemented testing for device_notice

iblanco11981870 committed:
Merge branch 'trunk-patch' into file_object_notice_stream

iblanco11981870 committed:
improve tests for notice

iblanco11981870 committed:
parameterized device notice tests

iblanco11981870 committed:
Update credits.rst

iblanco11981870 committed:
change documentation for notice change

iblanco11981870 committed:
Merge branch 'file_object_notice_stream' of https://github.com/glotzerlab/hoomd-blue into file_object_notice_stream

iblanco11981870 committed:
added the class noticefile

iblanco11981870 committed:
Merge branch 'file_object_notice_stream' of https://github.com/glotzerlab/hoomd-blue into file_object_notice_stream

iblanco11981870 committed:
implemented level in init

joaander commented on hoomd/device.py:
```diff
@@ -27,6 +27,42 @@
 
 
+class NoticeFile:
+    """Write a notice message to a file using `hoomd.Device`.
```
```suggestion
    """A file-like object that writes to a `hoomd.Device` notice stream.
```

joaander commented on hoomd/pytest/test_table.py:
Yes, please restore `test_table.py` and `table.py` to their previous state. You can clean up the history with an interactive rebase, or make a new branch and cherry-pick the changes to keep: c84ba40 and 312aa9c (plus any commits you make applying suggestions).

iblanco11981870 replied:
I just pushed a version that restores test_table.py and table.py to their previous state.

joaander commented on hoomd/device.py:
```diff
@@ -27,6 +27,42 @@
+            message (str): Message to write.
+        """
+        self._buff += str(message)
+        lines = self._buff.split("\n")
+        for line in lines:
+
+            if "\n" == line[-2:]:
+                pass
+            else:
+                line += "\n"
+
+            self._msg.notice(self._level, line)
+        self._buff = ""
```
This is not working as intended.
With this test script:
```
import hoomd
cpu = hoomd.device.CPU()
notice_file = hoomd.device.NoticeFile(cpu)

notice_file.write('t')
notice_file.write('e')
notice_file.write('s')
notice_file.write('t')
notice_file.write('1')
notice_file.write('\n')

notice_file.write('test2')
notice_file.write('\n')

notice_file.write('test3\ntest4\n')
```
I get the output:
```
t
e
s
t
1


test2


test3
test4

```
where I would expect:
```
test1
test2
test3
test4
```
(you can set `notice_file` to `sys.stdout` to confirm).

This change allows buffering of input missing `\n` over to the next call to `write` and avoids adding extra newlines to the output.
```suggestion
        lines = self._buff.split("\n")
        for line in lines[:-1]:
            self._msg.notice(self._level, line + "\n")

        if self._buff.endswith("\n"):
            self._msg_notice(self._level, lines[-1] + "\n")
            self._buff = ""
        else:
            self._buff = lines[-1]
```

joaander commented on hoomd/device.py:
```diff
@@ -27,6 +27,42 @@
+
+    Args:
+        device (`hoomd.Device`): The `Device` object.
+        level (int): Message notice level. Default value is 1
```
```suggestion
        level (int): Message notice level. Default value is 1.
```

joaander commented on hoomd/device.py:
```diff
@@ -27,6 +27,42 @@
+        self._buff = ""
+
+    def flush(self):
+        """Flush the device."""
```
```suggestion
        """Flush the output."""
```

joaander commented on hoomd/device.py:
```diff
@@ -27,6 +27,42 @@
+        self._level = level
+
+    def write(self, message):
+        """Writes the notice message to the device.
```
```suggestion
        """Writes data to the associated devices notice stream.
```

joaander commented on hoomd/device.py:
```diff
@@ -27,6 +27,42 @@
+    Args:
+        device (`hoomd.Device`): The `Device` object.
+        level (int): Message notice level. Default value is 1
+    """
```
```suggestion
    Note:
        Use this in combination with `Device.message_filename` to combine notice
        messages with output from code that expects file-like objects (such as
        `hoomd.write.Table`.
    """
```

joaander reviewed (changes requested):
Thanks! Let's make a few changes, then this is ready to merge and include in v4.0.

iblanco11981870 committed:
Update hoomd/device.py

Improve description

Co-authored-by: Joshua A. Anderson <joaander@umich.edu>

iblanco11981870 committed:
Update hoomd/device.py

Script takes into account buffering of input missing \n over to the next call to write and avoids adding extra new lines to the output.

Co-authored-by: Joshua A. Anderson <joaander@umich.edu>

iblanco11981870 committed:
Update hoomd/device.py

Improve documentation

Co-authored-by: Joshua A. Anderson <joaander@umich.edu>

iblanco11981870 committed:
Update hoomd/device.py

Improve documentation for flush

Co-authored-by: Joshua A. Anderson <joaander@umich.edu>

iblanco11981870 committed:
Update hoomd/device.py

Improve documentation to explain the utilization of hoomd.device.notice

Co-authored-by: Joshua A. Anderson <joaander@umich.edu>

iblanco11981870 committed:
Update hoomd/device.py

Improve documentation of the write function

Co-authored-by: Joshua A. Anderson <joaander@umich.edu>

iblanco11981870 committed:
reversed the changes in table.py and test_table.py

joaander commented:
@iblanco11981870 It appears that your force push reverted all of the suggested changes. Did you perhaps commit suggestions on GitHub, then force push from local?

iblanco11981870 commented:
> @iblanco11981870 It appears that your force push reverted all of the suggested changes. Did you perhaps commit suggestions on GitHub, then force push from local?

That is exactly what I did. Should I now do the suggested changes manually and add them to my local environment?

joaander commented on hoomd/pytest/test_device.py:
```diff
@@ -138,3 +138,36 @@ def test_device_notice(device, tmp_path):
+
+    # Message file declared. Should output in specified file.
+    device.message_filename = str(tmp_path / "str_message")
+    msg = "This message should output."
```
These tests are failing with the changes I suggested to `write`. You need to include `'\n'` in the messages to write. Also, it would be helpful if each `assert` as in a separate `test_noticefile_*(device, tmp_path)` function.

joaander reviewed (changes requested):
Thanks! Now the tests need to be updated to work with the fixed `write` method.

iblanco11981870 committed:
Merge branch 'trunk-patch' into file_object_notice_stream

iblanco11981870 committed:
fixed the test appropiately and also adapted the device.py code to accept integer outputs

joaander commented on hoomd/device.py:
```diff
@@ -27,6 +27,50 @@
+        Args:
+            message (str): Message to write.
+        """
+        if type(message) != str:
+            self._buff += str(message) + "\n"
+        else:
+            self._buff += message
```
```suggestion
        self._buff += message
```

file-like objects expect `str` arguments. It is the responsibility of the caller to convert non-string arguments into strings with appropriate formatting.
```
>>> import sys
>>> sys.stdout.write(1234)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: write() argument must be str, not int
```

You can add an explicit type check and `raise TypeError`, or let the `+=` operator on `str` raise the error. Then you should remove the test that attempts to write an int.

joaander commented on hoomd/pytest/test_device.py:
```diff
@@ -138,3 +138,37 @@ def test_device_notice(device, tmp_path):
+            assert fh.read() == test_msg + "\n"
+
+    # Test the level argument
+    msg = "This message should not output./n"
```
```suggestion
    msg = "This message should not output.\n"
```

joaander replied:
'\n' is a newline character. '/n' is the characters '/' and 'n'.

joaander commented on hoomd/pytest/test_device.py:
```diff
@@ -138,3 +138,37 @@ def test_device_notice(device, tmp_path):
+        with open(device.message_filename) as fh:
+            assert fh.read() == str(msg) + "\n"
+
+    # Test notice with a message that is not a string.
+    msg = 123456
+    test_msg = str(msg) + "\n"
+    device.message_filename = str(tmp_path / "int_message")
+    notice_file.write(msg)
+
+    if device.communicator.rank == 0:
+        with open(device.message_filename) as fh:
+            assert fh.read() == test_msg + "\n"
```
```suggestion
    # Test notice with a message that is not a string.
    msg = 123456
    device.message_filename = str(tmp_path / "int_message")
    with pytest.raises(TypeError):
         notice_file.write(msg)
```

joaander reviewed (changes requested):
There was no need to merge `trunk-patch` into this branch. The feature you are developing does not rely on any of those bug fixes. GitHub is now confused about which changes were actually made on your branch and is showing changes from other pull requests. To attempt to resolve this, merge the latest `trunk-major` into your branch.

iblanco11981870 committed:
merged with trunk-major

iblanco11981870 committed:
Update hoomd/device.py

Co-authored-by: Joshua A. Anderson <joaander@umich.edu>

iblanco11981870 committed:
Update hoomd/pytest/test_device.py

Co-authored-by: Joshua A. Anderson <joaander@umich.edu>

iblanco11981870 committed:
Update hoomd/pytest/test_device.py

Co-authored-by: Joshua A. Anderson <joaander@umich.edu>

iblanco11981870 commented:
> There was no need to merge `trunk-patch` into this branch. The feature you are developing does not rely on any of those bug fixes. GitHub is now confused about which changes were actually made on your branch and is showing changes from other pull requests. To attempt to resolve this, merge the latest `trunk-major` into your branch.

My apologies. I just did that. Please let me know if there is anything else I should do

joaander commented:
It appears that GitHub will remain hopelessly confused. Reviewers will need to ignore the changes GitHub is showing that come from #1551.

joaander commented:
pre-commit.ci autofix

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

joaander committed:
Merge branch 'trunk-major' into file_object_notice_stream

joaander committed:
Add NoticeFile to the documentation.

joaander reviewed (approved):
Thanks for working on this! I made a few changes so that `NoticeFile` will appear in the documentation. With that, this is good to merge.
